usagov/usagov-2021
# Usagov-1640 Wizard front and back end benchmark

## Jira Task

This PR is meant as a benchmark of the progress, not the finished product, so the epic is provided here as opposed to an individual ticket.
https://cm-jira.usa.gov/browse/USAGOV-412

## Description

Create new wizard vocabulary and associated pathauto pattern. Setup hierarchy_manager module to administer terms created in this vocabulary. Created wizard_options view to programmatically generate a list of children of the current taxonomy term indicated by route context. Referenced this view in a field added to the wizard vocabulary. Add link to a "wizard manager" under the usagov tools dropdown in the administration toolbar.  Setup menu_item_fields module to allow menu entities to become fieldable. Create a new custom module with a service that can check if a wizard taxonomy term is referenced in the left navigation menu. If it is, use the new menu_entity field to decide whether to use this link as a reference point to generate a left navigation menu specific to the wizard. Create custom preprocess logic for wizard breadcrumbs, wizard forms, and wizard pages. Create new twig templates for wizard taxonomy terms, wizard breadcrumbs, and the wizard options view. Create a new scss file for wizard front-end theming. Create two new wizard js files and attach them to the new twig templates. Remove the old wizard js files from global libraries and attach them to the twig for the wizard nodes. 

## Type of Changes

- [ X ] New Feature
- [ X ] Frontend (Twig, Sass, JS)
  ![image](https://github.com/usagov/usagov-2021/assets/96207895/b5a42f46-8c37-48bd-9d91-3ae06ca2d3cd)
- [ X ] Drupal Config (requires "drush cim")
- [ X ] New Modules (requires rebuild)

## Testing Instructions

#### Verify that old wizard is unaffected
1. Navigate to /where-report-scams.
2. Navigate through the wizard. Everything appears as expected.
3. Attempt to create new wizard and wizard step nodes and integrate them with the existing tool.
4. Everything works as expected

#### Test out the new wizard
Please note that this is not the finished product. The purpose of this PR is to get the bulk of the work merged in. So the testing instructions are not thorough at this point.

1. Find the "Wizard Manager" link contained in the USAGov tools dropdown and click on it.
2. Add terms based on the [figma](https://www.figma.com/design/sznp6ZqUSiuRBDYcoiL1PO/FY24-Scam-Wizard-Mobile%2BDesktop-Prototype?node-id=138-14497&t=9FarCvkN4gmcjxku-0), with the first term being the entry point, the page with the title "Learn where to report a scam".
3. Go to to the Left Menu English menu management page. Change the link for How to Report a Scam to direct it to /learn-where-report-scam (or whatever the alias ends up being for the entry level page in the wizard). Before saving, be sure to check the "custom parent" box.
4. Do the same thing for Left Menu Spanish
5. Continue to add terms until the whole wizard is accounted for.
6. Do the same thing, now in Spanish.
7. Test moving through the wizard. 
8. Navigate to /scams-and-fraud.
9. Your new wizard taxonomy term shows up in the list of cards, with the expected content.

As a tester, you are verifying after all of this that all of the major components of the wizard are present. You were able to create pages more or less that match the figma file. You were able to easily add children to each page, and manage the hierarchy of the wizard from the management screen. There are previous and next buttons on each page that go to the correct location. If you try to click next without selection an option, an error message appears. The "start over" button is present on pages starting with the 3rd level of the hierarchy and if clicked takes you back to the 2nd level of the hierarchy. The left navigation menu shows as expected, and the links in it are clickable. 
### Requires New Config
- [ X ] Yes

### Requires New Content
- [ X ] Yes

## Security Review


- [ X ] Adds/updates software (including a library or Drupal module)
- [ X ] Communication with external service
- [ X ] Changes permissions or workflow


## Reviewer Reminders
- Reviewed code changes
- Reviewed functionality
- Security review complete or not required

## Post PR Approval Instructions
Follow these steps as soon as you merge the new changes.

1. Go to the [USAGov Circle CI project](https://app.circleci.com/pipelines/github/usagov/usagov-2021).
10. Find the commit of this pull request.
11. Build and deploy the changes.
12. Update the Jira ticket by changing the ticket status to `Review in Test` and add a comment. State whether the change is already visible on [cms-dev.usa.gov](http://cms-dev.usa.gov/) and [beta-dev.usa.gov](http://beta-dev.usa.gov/), or if the deployment is still in process.
